rcl/test/rcl/test_graph.cpp

```diff
@@ -310,7 +313,218 @@ check_graph_state(
+  rcl_ret_t ret;
+  rcl_names_and_types_t nat{};
+  nat = rcl_get_zero_initialized_names_and_types();
+  ret = func(node, topic_name, &nat);
+  EXPECT_EQ(num_topics, nat.names.size);
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  rcl_reset_error();
+  ret = rcl_names_and_types_fini(&nat);
+  rcl_reset_error();
+}
+
+/**
+ * This test creates publishers, subscribers, and services, then calls node graph functions
+ * from the perspective of each node.
+ * It verifies each node perceives the same graph.
+ */
+TEST_F(CLASSNAME(TestGraphFixture, RMW_IMPLEMENTATION), test_node_info_functions) {
+  std::string topic_name("/test_node_info_functions__");
+  std::chrono::nanoseconds now = std::chrono::system_clock::now().time_since_epoch();
+  topic_name += std::to_string(now.count());
+  RCUTILS_LOG_INFO_NAMED(ROS_PACKAGE_NAME, "Using topic name: %s", topic_name.c_str());
+  rcl_ret_t ret;
+  const rcl_guard_condition_t *graph_guard_condition =
+    rcl_node_get_graph_guard_condition(this->node_ptr);
+
+  const rcl_guard_condition_t *remote_graph_guard_condition =
+    rcl_node_get_graph_guard_condition(this->node_ptr);
+
+  auto remote_node_ptr = new rcl_node_t;
+  *remote_node_ptr = rcl_get_zero_initialized_node();
+  const char *remote_node_name = "remote_graph_node";
+  rcl_node_options_t node_options = rcl_node_get_default_options();
+  ret = rcl_node_init(remote_node_ptr, remote_node_name, "", &node_options);
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+
+  // Now create a publisher on "topic_name" and check that it is seen.
+  rcl_publisher_t pub = rcl_get_zero_initialized_publisher();
+  rcl_publisher_options_t pub_ops = rcl_publisher_get_default_options();
+  auto ts = ROSIDL_GET_MSG_TYPE_SUPPORT(test_msgs, msg, Primitives);
+  ret = rcl_publisher_init(&pub, this->node_ptr, ts, topic_name.c_str(), &pub_ops);
+  EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  rcl_reset_error();
+
+  // Now create two subscribers.
+  rcl_subscription_t sub = rcl_get_zero_initialized_subscription();
+  rcl_subscription_options_t sub_ops = rcl_subscription_get_default_options();
+  ret = rcl_subscription_init(&sub, this->node_ptr, ts, topic_name.c_str(), &sub_ops);
+  EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  rcl_reset_error();
+
+  // Now create a subscriber.
+  rcl_subscription_t sub2 = rcl_get_zero_initialized_subscription();
+  rcl_subscription_options_t sub_ops2 = rcl_subscription_get_default_options();
+  ret = rcl_subscription_init(&sub2, remote_node_ptr, ts, topic_name.c_str(), &sub_ops2);
+  EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  rcl_reset_error();
+
+  const char * service_name = "test_service";
+  rcl_service_t service = rcl_get_zero_initialized_service();
+  rcl_service_options_t service_options = rcl_service_get_default_options();
+  auto ts1 = ROSIDL_GET_SRV_TYPE_SUPPORT(test_msgs, Primitives);
```

Comment:
This API is outdated.

```suggestion
  auto ts1 = ROSIDL_GET_SRV_TYPE_SUPPORT(test_msgs, srv, Primitives);
```

Reply:
Sounds good!